external/fv3fit/fv3fit/keras/_models/recurrent.py

```diff
@@ -0,0 +1,611 @@
+            outputs=tendency_outputs,
+        )
+
+        input_layers, forcing_input = get_vector(
+            self.input_packer, self.input_scaler, series=False
+        )
+        state_layers, state_input = get_vector(
+            self.prognostic_packer, self.prognostic_scaler, series=False
+        )
+        denormalized_state = self.prognostic_packer.pack_layer()(state_layers)
+        x = tf.keras.layers.concatenate([forcing_input, state_input])
+        predicted_tendency = get_predicted_tendency(x, denormalized_state)
+
+        tendency_outputs = self.prognostic_packer.unpack_layer(feature_dim=1)(
+            predicted_tendency
+        )
+
+        predict_keras_model = tf.keras.Model(
+            inputs=input_layers + state_layers, outputs=tendency_outputs
+        )
+
+        train_keras_model.compile(optimizer=self.optimizer, loss=self.losses)
+
+        return train_keras_model, train_tendency_keras_model, predict_keras_model
+
+    def fit(self, X: xr.Dataset, *, epochs=1):
+        """
+        Args:
+            X: training data
+            epochs: how many epochs to train. default is 1 for the case when X
+                contains one batch at a time
+        """
+        if self.train_keras_model is None:
+            raise RuntimeError("Must call `build_for` method before calling fit")
+        if tuple(self.prognostic_packer.pack_names) != (
+            "air_temperature",
+            "specific_humidity",
+        ):
+            raise ValueError(
+                "given tendency is hard-coded to use air_temperature and "
+                "specific_humidity as prognostic names, if the prognostic "
+                "packer does not use them there must be a bug "
+                "or this check should be removed. "
+                f"got names {self.prognostic_packer.pack_names}"
+            )
+
+        for name in list(self.input_packer.pack_names) + list(
+            self.prognostic_packer.pack_names
+        ):
+            assert X[name].dims[1] == BPTTModel.TIME_DIM_NAME
+
+        self.train_keras_model.fit(
+            x=self.get_keras_inputs(X),
+            y=self.get_target_state(X),
+            batch_size=self.train_batch_size,
+            epochs=epochs,
+            shuffle=True,
+        )
+
+    def loss(self, X: xr.Dataset):
+        """
+        Return the loss on a sequence dataset.
+
+        Useful for getting validation loss.
+        """
+        if self.train_keras_model is None:
+            raise RuntimeError("must build model before loss can be calculated")
+        else:
+            output = self.train_keras_model.predict(self.get_keras_inputs(X))
+            target = self.get_target_state(X)
+            loss = sum(
+                np.mean(f(truth, pred))
+                for f, truth, pred in zip(self.losses, target, output)
+            )
+            return loss
+
+    def get_keras_inputs(self, X) -> Sequence[np.ndarray]:
```

Comment:
I think I can -- I think this is only called from unit tests.

Reply:
I'm going to leave this as public now because this functionality should really be package-private, and the class it lives on is already package-private. It's necessary to use the logic in this routine to interact with any of the `tf.keras.Model` objects on it, which is also currently package private.

If you like I could make all of this object-private instead (including the tensorflow models), it's not a super important distinction from my end. Mostly this was very useful to be able to use from `train_bptt` during development.